text/0784-no-globals.md

```diff
@@ -0,0 +1,494 @@
+        module.exports = (function() {
+-          var d = _eai_d;
++          var d = globalThis.__ember_12345_define;
+-          var r = _eai_r;
++          var r = globalThis.__ember_12345_require;
+          window.emberAutoImportDynamic = function(specifier) {
+            if (arguments.length === 1) {
+              return r("_eai_dyn_" + specifier);
+            } else {
+              return r("_eai_dynt_" + specifier)(Array.prototype.slice.call(arguments, 1));
+            }
+          };
+          window.emberAutoImportSync = function(specifier) {return r("_eai_sync_" + specifier)(Array.prototype.slice.call(arguments, 1));};
+          d("lodash-es/find", [], function() { return __webpack_require__(/*! lodash-es/find */ "./node_modules/lodash-es/find.js");});
+          d("lodash-es/startsWith", [], function() {return __webpack_require__(/*! lodash-es/startsWith */ "./node_modules/lodash-es/startsWith.js");});
+          d("_eai_dyn_lodash-es/concat", [], function() {return __webpack_require__.e(/*! import() */ "node_modules_lodash-es_concat_js").then(__webpack_require__.bind(__webpack_require__, /*! lodash-es/concat */ "./node_modules/lodash-es/concat.js"));});
+        })();
+      }),
+
+-    "../../../../private/var/folders/q3/yzggf_l965zbzps_bk8xr37c0000gn/T/broccoli-4171121iPrUeBlpJ8/cache-307-webpack_bundler_ember_auto_import_webpack/l.js":
+-      (function(module, exports) {
+-        window._eai_r = require;
+-        window._eai_d = define;
+-      })
+  });
+  // ... webpack internals 
+})();
+```
+</details>
+
+> Note that ember-auto-import needs knowledge about the obfuscated names of the (still globally defined) functions (e.g.
+> `__ember_12345_define` in this example) here. This should be exposed to privileged addons like ember-auto-import or
+> Embroider *somehow*, but as this is a strictly private API, it is not detailed here.
+
+</details>
+
+---
+
+#### EmberENV
+
+Ember reads environment variables from a `EmberENV` hash to enable/disable certain runtime features. Usually this is set
+from the app's `config/environment.js` and `@ember/optional-features`, and compiled into the top of `vendor.js`:
+
+<details>
+<summary>vendor.js</summary>
+
+```js
+window.EmberENV = (function(EmberENV, extra) {
+  for (var key in extra) {
+    EmberENV[key] = extra[key];
+  }
+
+  return EmberENV;
+})(window.EmberENV || {}, {"FEATURES":{},"EXTEND_PROTOTYPES":{"Date":false},"_APPLICATION_TEMPLATE_WRAPPER":false,"_DEFAULT_ASYNC_OBSERVERS":true,"_JQUERY_INTEGRATION":false,"_TEMPLATE_ONLY_GLIMMER_COMPONENTS":true});
+```
+</details>
+
+But as you can see there, you can also inject those by setting a globally defined `window.EmberENV`.
+
+In `no-globals` mode though, the latter behavior will be disabled. Only the variables built into `vendor.js` are taken
+into account. Implementation wise this can happen in the same way as described above, by defining `EmberENV` within the
+IIFE (and not using the explicit `window.` namespace).
```

Comment:
I think it would be better to have Ember leverage `@embroider/macros` `getConfig` or `getOwnConfig` inside Ember's own code where we currently read `window.EmberENV` instead of leveraging the IIFE style.

Reply:
It seems fine to still support _reading_ from `window.EmberENV` even in no-globals mode? Otherwise we need to propose _something_ for the Ember Inspector to [enable the debug render tree](https://github.com/emberjs/ember-inspector/blob/15f49e5ab04e738ec540550001ad8a1b1537eb70/skeletons/web-extension/boot.js#L13).

Reply:
Actually, come to think about it, I think the inspector in its current form/architecture will most likely just completely break in no-globals mode as it relies on being able to get access to the things you are trying to no leak with this RFC. Probably not a good enough reason to hold this back, but it's definitely a drawback that should be noted. We can overhaul the inspector in theory but it will require more design than is in this RFC currently, and it's probably not fair to have this RFC solve that problem, plus I am not sure if anyone has time to work on a big inspector project. In the short/medium term, I think the inspector will probably just not work with `no-globals` mode. Is that an acceptable drawback?

Reply:
(cc @rwwagner90)

Reply:
Oh, that's a good point! I didn't think about that. Certainly needs to be mentioned. I don't know enough about the Inspector to assess how much needs to change to support `no-globals`. I guess all those possible ways to coordinate between those privileged tools what the scoped version of `require()` & Co are (e.g. `getConfig`) will not work for the Inspector, as it is a standalone app that does not take part in the build system of the *app*...

Maybe there still could be some escape hatch like defining `window.__ember_scoped_globals_dont_ever_use_this_in_user_code = 'myEmberApp_somerandomhash'`, when the app has `require` (and `define`) scoped as window.myEmberApp_somerandomhash.require`?

> Is that an acceptable drawback?

I would still say yes. You can still define the flag conditionally, like enable it for test and prod, while disabling it for dev!